lib/ruby_smb/gss/provider/ntlm.rb

```diff
@@ -0,0 +1,337 @@
+module RubySMB
+  module Gss
+    module Provider
+      #
+      # A GSS provider that authenticates clients via the NT LAN Manager (NTLM) Security Support Provider (NTLMSSP)
+      # protocol.
+      #
+      class NTLM < Base
+        # An account representing an identity for which this provider will accept authentication attempts.
+        Account = Struct.new(:username, :password, :domain) do
+          def to_s
+            "#{domain}\\#{username}"
+          end
+        end
+
+        class OSVersion < BinData::Record
+          endian :big
+
+          uint8  :major
+          uint8  :minor
+          uint16 :build
+          uint32 :ntlm_revision, initial_value: 15
+
+          def to_s
+            "Version #{major}.#{minor} (Build #{build}); NTLM Current Revision #{ntlm_revision}"
+          end
+        end
+
+        class Authenticator < Authenticator::Base
+          NEGOTIATE_FLAGS = {
```

Comment:
I actually was using the `Net::NTLM` flag definitions when I first started by found them difficult to use while debugging for two reasons.
1. Using the hex values made it difficult to understand what bit field each corresponded to.
2. As you noticed, the flag names are different. I was using Wireshark to debug some issues and some of the names were different and others were missing. This made for a cumbersome experience. Take `MBZ7` for example, I have no idea what that is but it corresponds to Anonymous in Wireshark.

Reply:
Absolutely, there are a few undocumented/unknown flags in the `Net::NTLM` library, including the Anonymous flag. I fixed them in the [PR](https://github.com/WinRb/rubyntlm/pull/38/files#diff-d0a5655805436a73145e2fed15c8916c907f9874e89670269615d015041c54bcR16) I submitted. It is based on this [documentation](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nlmp/99d90ff4-957f-4c8a-80e4-5bfe5a9a9832?redirectedfrom=MSDN).

That said, I agree that the bit field matches the documentation, so we can use your implementation. Would it be possible to move them in a more generic location and update the client to use them instead of `Net::NTLM` flags? I believe there is just one place where these flags are used: https://github.com/rapid7/ruby_smb/blob/master/lib/ruby_smb/client.rb#L652-L660

Reply:
Fixed in commit c3894902521bdd29240dc5a60bf220ea209d3583. I also moved the `OSVersion` struct since that's a field in some of the NTLM messages and could be used elsewhere as well.